storage/src/adb/immutable.rs

```diff
@@ -337,9 +377,14 @@ impl<E: RStorage + Clock + Metrics, K: Array, V: Codec, H: CHasher, T: Translato
         Ok(None)
     }
 
-    /// Get the value of the operation with location `loc` in the db if it matches `key`. The
-    /// location is assumed valid and this function panics otherwise.
+    /// Get the value of the operation with location `loc` in the db if it matches `key`. Returns
+    /// OperationPruned error if loc precedes the oldest retained location. The location is
+    /// otherwise assumed valid.
     pub async fn get_from_loc(&self, key: &K, loc: u64) -> Result<Option<V>, Error> {
+        if loc < self.oldest_retained_loc {
+            return Err(Error::OperationPruned(loc));
```

Comment:
I wonder if it is better just to return `Ok(None)`?

I suppose we "know we had it" given the current structure of this DB but if we allow state sync that will no longer be guaranteed. 🤔

Reply:
I suppose you ever encountering this error requires you to have called prune at some index greater than or equal to loc.

I looked at journal and saw that we return an error if you try to read from a pruned section (so this behavior is consistent): https://github.com/commonwarexyz/monorepo/blob/31c6abdf1a264beb727817cdd20b34295b3a3fe4/storage/src/journal/variable.rs#L222